alacritty_terminal/src/term/mod.rs

```diff
@@ -1826,24 +1812,29 @@ impl<T: EventListener> Handler for Term<T> {
 
         if let Ok(bytes) = base64::decode(base64) {
             if let Ok(text) = str::from_utf8(&bytes) {
-                self.clipboard.store(clipboard_type, text);
+                self.event_proxy.send_event(Event::ClipboardStore(clipboard_type, text.to_owned()));
             }
         }
     }
 
-    /// Write clipboard data to child.
+    /// Request to paste to a child.
     #[inline]
-    fn write_clipboard<W: io::Write>(&mut self, clipboard: u8, writer: &mut W, terminator: &str) {
+    fn request_clipboard_write(&mut self, clipboard: u8, terminator: &str) {
         let clipboard_type = match clipboard {
             b'c' => ClipboardType::Clipboard,
             b'p' | b's' => ClipboardType::Selection,
             _ => return,
         };
 
-        let text = self.clipboard.load(clipboard_type);
-        let base64 = base64::encode(&text);
-        let escape = format!("\x1b]52;{};{}{}", clipboard as char, base64, terminator);
-        let _ = writer.write_all(escape.as_bytes());
+        let terminator = terminator.to_owned();
+
+        self.event_proxy.send_event(Event::ClipboardLoad(
+            clipboard_type,
+            std::sync::Arc::new(move |text| {
+                let base64 = base64::encode(&text);
+                format!("\x1b]52;{};{}{}", clipboard as char, base64, terminator)
+            }),
```

Comment:
Lovely, this is exactly what I was thinking. Except for the Arc this looks like a relatively clean solution.